### ISSUE TITLE
Adjusting Training Scheme

### ISSUE THREAD

**Issue by Zador-Pataki** [facebookresearch/silk#24]
Hi, thanks for the great work. I would like to make a number of adjustments to the training script. 
1. load image pairs and specify correspondences. I would like correspondence masking, i.e. providing correspondences only for a subset of the available pixels. 
2. use a completely separate validation script during training 
3. adjust the loss function

If you could share with me minimal instructions on where and how to make these adjustments it would be greatly appreciated!

**Comment by gleize** [comment]
Hi @Zador-Pataki,

> 1. load image pairs and specify correspondences. I would like correspondence masking, i.e. providing correspondences only for a subset of the available pixels.
> 2. use a completely separate validation script during training
> 3. adjust the loss function

This can be done, but requires a fair amount of work. I can list a few pointers to help you with the change.
The easiest and quickest way would be to copy or override existing files :
* `etc/mode/train-silk.yaml` : This file constains dataset, dataloader and collate function information, as well as a pointer to which model to use (e.g. [here](https://github.com/facebookresearch/silk/blob/b601cbc5458c0e5504e75bff4e138532c1ce683a/etc/mode/train-silk.yaml#L3)). This is considered the main entrypoint config file (i.e. the selected file when calling `./bin/silk-cli mode=train-silk`).
  In this file, you will have to point to your new dataset (c.f. [here](https://github.com/facebookresearch/silk/blob/b601cbc5458c0e5504e75bff4e138532c1ce683a/doc/usage/dataset.md) to see how existing datasets are configured). And you will have to change the collate function. To make it easy, you can simply point to a custom function of yours by changing this [line](https://github.com/facebookresearch/silk/blob/b601cbc5458c0e5504e75bff4e138532c1ce683a/etc/mode/train-silk.yaml#L63) to something like this :
```yaml
collate_fn:
  _target_: your_module.your_custom_collate_fn
```
and remove everything below.
Your collate function will be fed to the dataloader and works as specified in [DataLoader](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader). So, your newly added dataset will have to provide pairs of images, the correspondences and the masks, which will be then fed to the collate function for batching.
* `lib/models/silk.py` : This file contains the training flow. The output of your collate function defined above will be given as the `batch` parameter to the `training_step` and `validation_step` methods (c.f. [here](https://github.com/facebookresearch/silk/blob/b601cbc5458c0e5504e75bff4e138532c1ce683a/lib/models/silk.py#L338)). In those functions, you can implement whatever you need (it will be easier in your case since you require a fair amount of change in correspondences, validation and loss function). You might want to re-use a few of our functions though so I'm listing them here :
  * [Image Augmentation](https://github.com/facebookresearch/silk/blob/b601cbc5458c0e5504e75bff4e138532c1ce683a/lib/models/silk.py#LL283C9-L283C20).
  * Calling the model backbone : 
  ```python
  desc, kp_logits = self._model.forward_flow(output=("normalized_descriptors", "logits"), images=images)
  ```
  * Calling the existing loss :
  ```python
  desc_loss, kp_loss, _, _ = self._loss(
            descriptors_0,
            descriptors_1,
            corr_forward,
            corr_backward,
            logits_0,
            logits_1,
        )
  ```
  
  That should give you a good starting point. Don't hesitate to ask if you need more info / clarifications.

**Comment by Zador-Pataki** [comment]
I have checked out everything you have shared and this is super helpful. Closing issue for now as I do not anticipate I will need more help. Thanks a lot!

**Comment by Zador-Pataki** [comment]
Hi, I have successfully adapted the pipeline, thanks! However, I have decided to stick to most of your implemented functions and make only a few low-level adaptations in the `training_step`. I am adapting some of the functions that you have added to your flow `self.flow.define_transition(...)`. In particular, I am changing `def _cet_corr(self, ...)` in `class SiLKRandomHomographies(...)`. 

In my setup, I am feeding the function correspondences from reference to query image, instead of feeding the Homographic samplers. I understand that the correspondences I am providing the function are equivalent to the `warped_positions_forward`, and `warped_positions_backward` (also I need to filter out the correspondences near the boundaries, where the pixels do not have extracted descriptors). 

My only question is about the function `positions_to_unidirectional_correspondence(...)`. Can I comfortably apply this function to whatever my correspondences are? Based on the paper and the function in [loss.py](https://github.com/facebookresearch/silk/blob/main/lib/losses/info_nce/loss.py) it seems like it makes the correspondences between discrete pixels instead of float values, and converts it into a different form, however, it is wrapped in something I don't completely understand [jax2torch(...)](https://github.com/facebookresearch/silk/blob/main/lib/losses/info_nce/__init__.py).

I hope you can give some quick insight if there is anything that I should be careful of in this implementation. Otherwise, the code is very well organized and with some effort was easy to interpret!

**Comment by gleize** [comment]
Hi @Zador-Pataki,

> [...] about the function positions_to_unidirectional_correspondence(...). Can I comfortably apply this function to whatever my correspondences are?

Yes. `jax2torch(...)` is a wrapper that enables us to call jax functions from pytorch (with handling of brackpropagation when necessary). So, if your inputs to `positions_to_unidirectional_correspondence(...)` are similar to `warped_positions_forward/backward`, there shouldn't be any issues.

> Otherwise, the code is very well organized and with some effort was easy to interpret!

Thanks 😄

**Comment by Zador-Pataki** [comment]
Thanks! In the end, I had some trouble; for some reason using `.reshape` on my custom correspondences array (my alternative to the meshgrid) wasn't resulting in the appropriate order of inputs in `warped_positions_forward/backward` but anyway I found a workaround. Good to go!
